File: stage-8/part-2/answer.py
import math
import os

class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.size = [1] * n
        self.components = n
    
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def union(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)
        
        if root_x == root_y:
            return False
        
        if self.size[root_x] < self.size[root_y]:
            self.parent[root_x] = root_y
            self.size[root_y] += self.size[root_x]
        else:
            self.parent[root_y] = root_x
            self.size[root_x] += self.size[root_y]
        
        self.components -= 1
        return True
    
    def is_connected(self):
        return self.components == 1

def parse_input(input_path):
    with open(input_path, 'r') as f:
        lines = f.read().strip().split('\n')
    points = []
    for line in lines:
        x, y, z = map(int, line.split(','))
        points.append((x, y, z))
    return points

def calculate_distance(p1, p2):
    dx = p1[0] - p2[0]
    dy = p1[1] - p2[1]
    dz = p1[2] - p2[2]
    return math.sqrt(dx * dx + dy * dy + dz * dz)

def solve(points):
    connections = []
    
    for i in range(len(points)):
        for j in range(i + 1, len(points)):
            distance = calculate_distance(points[i], points[j])
            connections.append((distance, i, j))
    
    connections.sort(key=lambda x: x[0])
    
    uf = UnionFind(len(points))
    last_connection = None
    
    for distance, i, j in connections:
        if uf.union(i, j):
            last_connection = (points[i][0], points[j][0])
            if uf.is_connected():
                break
    
    if last_connection:
        return last_connection[0] * last_connection[1]
    return 0

example_path = os.path.join(os.path.dirname(__file__), '../input.example')
real_path = os.path.join(os.path.dirname(__file__), '../input.real')

example_points = parse_input(example_path)
example_result = solve(example_points)
print(f'Example result: {example_result}')

if example_result != 25272:
    print(f'Example validation failed: expected 25272, got {example_result}')
    exit(1)

real_points = parse_input(real_path)
real_result = solve(real_points)
print(f'Real answer: {real_result}')

File: stage-8/part-1/answer.js
const fs = require('fs');
const path = require('path');
class UnionFind {
    constructor(n) {
        this.parent = Array.from({ length: n }, (_, i) => i);
        this.size = Array(n).fill(1);
    }
    find(x) {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]);
        }
        return this.parent[x];
    }
    union(x, y) {
        const rootX = this.find(x);
        const rootY = this.find(y);
        if (rootX === rootY) {
            return false;
        }
        if (this.size[rootX] < this.size[rootY]) {
            this.parent[rootX] = rootY;
            this.size[rootY] += this.size[rootX];
        }
        else {
            this.parent[rootY] = rootX;
            this.size[rootX] += this.size[rootY];
        }
        return true;
    }
    getSize(x) {
        return this.size[this.find(x)];
    }
    getCircuitSizes() {
        const rootSizes = new Map();
        for (let i = 0; i < this.parent.length; i++) {
            const root = this.find(i);
            rootSizes.set(root, this.size[root]);
        }
        return Array.from(rootSizes.values()).sort((a, b) => b - a);
    }
}
function parseInput(inputPath) {
    const content = fs.readFileSync(inputPath, 'utf-8');
    const lines = content.trim().split('\n');
    return lines.map((line) => {
        const [x, y, z] = line.split(',').map(Number);
        return { x, y, z };
    });
}
function calculateDistance(p1, p2) {
    const dx = p1.x - p2.x;
    const dy = p1.y - p2.y;
    const dz = p1.z - p2.z;
    return Math.sqrt(dx * dx + dy * dy + dz * dz);
}
function solve(points, connectionsToMake) {
    const connections = [];
    for (let i = 0; i < points.length; i++) {
        for (let j = i + 1; j < points.length; j++) {
            const distance = calculateDistance(points[i], points[j]);
            connections.push({ from: i, to: j, distance });
        }
    }
    connections.sort((a, b) => a.distance - b.distance);
    const uf = new UnionFind(points.length);
    let connectionsMade = 0;
    for (const connection of connections) {
        if (connectionsMade >= connectionsToMake) {
            break;
        }
        if (uf.union(connection.from, connection.to)) {
            connectionsMade++;
        }
    }
    const circuitSizes = uf.getCircuitSizes();
    const threeLargest = circuitSizes.slice(0, 3);
    return threeLargest.reduce((product, size) => product * size, 1);
}
const examplePath = path.join(__dirname, '../input.example');
const realPath = path.join(__dirname, '../input.real');
const examplePoints = parseInput(examplePath);
const exampleResult = solve(examplePoints, 10);
console.log(`Example result: ${exampleResult}`);
if (exampleResult !== 40) {
    console.error(`Example validation failed: expected 40, got ${exampleResult}`);
    process.exit(1);
}
const realPoints = parseInput(realPath);
const realResult = solve(realPoints, 1000);
console.log(`Real answer: ${realResult}`);
